### PR TITLE
[indexer] - add store procedure and table trigger to populate objects_history and owner tables

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

* Added store procedure and table trigger to populate objects_history, owner and owner_history tables
* objects table now stores only the latest objects
* object_history table stores all the changes to the objects, this table grows quickly so it is partitioned by epoch.
* owner table is now populated trigged by the objects table, it contain latest object owner index.
* owner_history table stores the changes of object ownership, triggered by owner table, it only record new, delete and owner change, object change without owner change will not be recorded.
* also added base58digest and address database types to make the table definition more consistent.

TODOs:
implement paginated get_owner_objects endpoint using the owner and owner_history table.
implement get_owner_objects_update endpoint using owner history table.
store bcs bytes in db for object and transaction table, instead of parsed json.
implement module cache in indexer for resolving bcs move objects.
implement epoch table, trigger partition creation using the epoch table

## Test Plan 

Currently we can only manually test this locally, will investigate if we can use docker to test the db in CI.

Screenshots of testing locally:
1. populated owner table
image

2. populated object_history table
image

3. populated owner_history table
image

### CODE REVIEW

File: crates/sui-indexer/migrations/2022-11-28-204251_addresses/up.sql [hunk 1]
```diff
@@ -1,7 +1,12 @@
+-- SuiAddress and ObjectId type, 0x + 64 chars hex string
+CREATE DOMAIN address VARCHAR(66);
+-- Max char length for base58 encoded digest
+CREATE DOMAIN base58digest VARCHAR(44);
```
Comment (666lcz) [inline crates/sui-indexer/migrations/2022-11-28-204251_addresses/up.sql:4]: I am not sure if it is a good idea to hardcode the lengths. Are we certain that these will never change again?
cc @joyqvq @kchalkias @lxfind

File: crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql [hunk 2]
```diff
@@ -1,32 +1,99 @@
...
+    owner_address          address,
+    initial_shared_version BIGINT,
+    previous_transaction   base58digest NOT NULL,
+    package_id             address      NOT NULL,
+    transaction_module     VARCHAR      NOT NULL,
+    object_type            VARCHAR      NOT NULL,
+    object_status          change_type  NOT NULL,
+    CONSTRAINT objects_history_pk PRIMARY KEY (epoch, object_id, version)
+) PARTITION BY RANGE (epoch);
+-- TODO: Add trigger to automatically create partitions at new epoch when we have Epoch Table
```
Comment (666lcz) [inline crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql:43]: Is my understanding correct that this needs to be done before the next release?

File: crates/sui-json-rpc-types/src/sui_object.rs [hunk 3]
```diff
@@ -48,8 +48,10 @@ pub struct SuiObjectData {
     /// Base64 string representing the object digest
     pub digest: ObjectDigest,
     /// The type of the object. Default to be None unless SuiObjectDataOptions.showType is set to true
+    #[schemars(with = "Option<String>")]
+    #[serde_as(as = "Option<DisplayFromStr>")]
```
Comment (666lcz) [inline crates/sui-json-rpc-types/src/sui_object.rs:52]: TIL. I was trying to use `ObjectType` but encountered some errors which could have been resolved by these two macros

File: crates/sui-indexer/src/models/objects.rs [hunk 4]
```diff
@@ -5,39 +5,51 @@ use crate::models::owners::OwnerType;
...
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
 use sui_types::object::Owner;
 
 #[derive(Queryable, Insertable, Debug, Identifiable, Clone)]
 #[diesel(primary_key(object_id))]
 pub struct Object {
-    #[diesel(deserialize_as = i64)]
-    pub id: Option<i64>,
+    pub epoch: i64,
+    pub checkpoint: i64,
```
Comment (666lcz) [inline crates/sui-indexer/src/models/objects.rs:16]: Nit: Let's add some comments like this is the epoch/checkpoint in which the (object_id, version) gets updated?

File: crates/sui-indexer/migrations/2023-03-01-193733_owner/up.sql [hunk 5]
```diff
@@ -1,36 +1,133 @@
...
+$body$
+    LANGUAGE plpgsql;
+
+CREATE TRIGGER owner
+    AFTER INSERT OR UPDATE OR DELETE
+    ON objects
+    FOR EACH ROW
+EXECUTE PROCEDURE owner_modified_func();
+
+CREATE TABLE owner_history
```
Comment (666lcz) [inline crates/sui-indexer/migrations/2023-03-01-193733_owner/up.sql:64]: Do you mind giving an example about how a query looks like for the `owner_history` table? How does the caller know which partition/epoch to fetch?

File: crates/sui-indexer/migrations/2022-11-28-204251_addresses/up.sql (hunk 1 above)
Comment (patrickkuo) [reply crates/sui-indexer/migrations/2022-11-28-204251_addresses/up.sql:4]: I don't think this can change again :| , however event if the key length change we can alter the domain really easily in DB

File: crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql (hunk 2 above)
Comment (gegaowp) [reply crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql:43]: to double check, this table can be used for txn query via mutated objects, right?

File: crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql [hunk 6]
```diff
@@ -1,32 +1,99 @@
...
-    package_id TEXT NOT NULL,
-    transaction_module TEXT NOT NULL,
-    object_type TEXT,
-
-    -- status can be CREATED, MUTATED or DELETED.
-    object_status VARCHAR(255) NOT NULL
+    previous_transaction   base58digest NOT NULL,
+    package_id             address      NOT NULL,
+    transaction_module     VARCHAR      NOT NULL,
+    object_type            VARCHAR      NOT NULL
```
[the remark this answers is not in the document]
Comment (patrickkuo) [reply crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql]: yes that's the plan, we also need to implement module cache in the indexer as well.

File: crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql (hunk 2 above)
Comment (patrickkuo) [reply crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql:40]: ya this table can be use like ordinary table, e.g. you can query the change history of object using 
```
SELECT checkpoint, version, object_digest, previous_transaction, object_status
FROM objects_history
WHERE object_id = '0x32f7eb342bd16ab03270544bd82dfae52699955a6c01c0560a5a38608a34de1c'
```
result:
image

Comment (patrickkuo) [reply crates/sui-indexer/migrations/2022-12-01-034426_objects/up.sql:40]: yes, this is taken from `sui_types::Object::pervious_transaction`

File: crates/sui-indexer/migrations/2023-03-01-193733_owner/up.sql (hunk 5 above)
Comment (patrickkuo) [reply crates/sui-indexer/migrations/2023-03-01-193733_owner/up.sql:72]: thanks to this not so noob question, make me realise this PR was missing the deleted objects handling. I have since added that back so the owner_history query will now.

to query the "current" object ownership, we can simply query the `owner` table, which contain the latest owner index.
e.g.
```
SELECT *
FROM owner
WHERE owner_type = 'address_owner'
  AND owner_address =
      '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a'
  AND object_status NOT IN ('deleted', 'wrapped', 'unwrapped_then_deleted');
```

to query the owner_history table, it is a bit more involved:
```
SELECT *
FROM (SELECT *,
             RANK() OVER (PARTITION BY object_id ORDER BY version DESC) version_rank
      FROM owner_history
      WHERE ((owner_type = 'address_owner' AND
              owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a')
          OR (old_owner_type = 'address_owner' AND
              old_owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a'))
        AND checkpoint <= 20) AS diff
WHERE version_rank = 1
  AND owner_address = '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a';
```
this query select all owner changes (gain and loss ownership) from the history table for a address before checkpoint 20, and then for each object in the subquery, select the highest version number(using `RANK() OVER`), and then select the one belongs to the address. 

in simple terms this query will give you all object owned by that address at a checkpoint.

This is not easy to query so I have created a DB function to make it easier:
```
SELECT *
FROM object_owned_at_checkpoint(20, 'address_owner',
                                '0xb6a6dc32eb5833b4ef95407f5dd4a3abfd0042dfc619091bfc7335fb2864f41a');
```

**Review by gegaowp — APPROVED** [review]
Looks great!

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 7]
```diff
@@ -492,13 +489,37 @@ impl IndexerStore for PgIndexerStore {
...
+                        .execute(conn)?;
+
+                    diesel::insert_into(objects::table)
+                        .values(&changes.deleted_objects)
+                        .on_conflict(objects::object_id)
+                        .do_update()
+                        .set((
+                            objects::epoch.eq(excluded(objects::epoch)),
+                            objects::checkpoint.eq(excluded(objects::checkpoint)),
+                            objects::version.eq(excluded(objects::version)),
```
[the remark this answers is not in the document]
Comment (patrickkuo) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:517]: this is mainly because we use trigger to populate the owner tables, with insert, we the trigger will receive the old and new data, which we will need to know at which checkpoint the delete was happened, if we just do delete, we won't be able to pass the checkpoint info to the delete statement.

we can delete the "deleted" object part of the archiving job later.
